

<div style="position: relative; background: linear-gradient(135deg, #24398A 0%, #1a2a66 100%); border-radius: 20px 20px 0px 0px; padding: 30px; box-shadow: 0 8px 16px rgba(36, 57, 138, 0.3);">
  
  <!-- Logo UNISON - Derecha -->
  <div style="position: absolute; top: 20px; right: 20px; background: white; padding: 4px; border-radius: 8px; box-shadow: 0 2px 4px rgba(0,0,0,0.2);">
    <img src="images/logo_unison.jpg" alt="UNISON" style="height: 160px;">
  </div>
  
  <!-- Imagen Izquierda -->
  <div style="position: absolute; top: 20px; left: 20px; background: white; padding: 4px; border-radius: 8px; box-shadow: 0 2px 4px rgba(0,0,0,0.2);">
    <img src="images/genetic_banner.jpg" alt="Curso" style="height: 160px;">
  </div>
   
  <!-- Título centrado -->
  <div style="color: white; max-width: 60%; margin: 0 auto; text-align: center;">
    <h1 style="color: #EBA93B; margin: 0; font-size: 28px; text-shadow: 2px 2px 4px rgba(0,0,0,0.3);">
      Metaheurísticas para Ciencia de Datos: <br>Teoría y Práctica
    </h1><br>
    <h3 style="margin: 15px 0 5px 0; font-size: 16px; opacity: 0.95;">Maestría en Ciencia de Datos</br></br>
    Ramón Soto C. / ramon.soto@unison.mx</h3>
  </div>
</div>

<div style="background: white; border: 3px solid #EBA93B; border-radius: 0px 0px 20px 20px; padding: 25px;">
  <div style="display: flex; align-items: center; margin-bottom: 15px;">
    <div style="background: #24398A; color: white; font-size: 24px; font-weight: bold; padding: 10px 20px; border-radius: 8px; margin-right: 20px;">05</div>
    <div>
      <h2 style="color: #24398A; margin: 0;">Sistemas Multiagente (SMA) para Ciencia de Datos</h2>
      <p style="color: #666; margin: 5px 0 0 0; font-style: italic;">Arquitecturas distribuidas para problemas complejos y no estacionarios</p>
    </div>
  </div>
  
  <div style="background: #f8f9fa; padding: 15px; border-radius: 5px; border-left: 4px solid #24398A;">
    <h4><strong>Presentación general:</strong></h4>
    <p>
      ¿Qué ocurre cuando ningún modelo individual puede capturar la complejidad del problema? Esta libreta propone un cambio de paradigma: de algoritmos individuales a sociedades de agentes que interactúan. Los Sistemas Multiagente (SMA) emergen como marco unificador que reinterpreta metaheurísticas como sociedades computacionales donde la inteligencia no reside en ningún componente individual, sino que emerge de sus interacciones. Veremos cómo PSO, GA, ACO y otras metaheurísticas pueden entenderse como SMA especializados, y exploraremos arquitecturas híbridas que combinan agentes reactivos, deliberativos y cognitivos. Este enfoque es particularmente valioso para problemas con información distribuida, no estacionariedad y múltiples objetivos en conflicto.
</p>
  </div>
</div>



<div style="padding: 15px 20px; background-color: #f8f9fa; margin: 20px 0;">
    <p style="font-size: 16px; line-height: 1.6; color: #333; margin: 0;">
        En dominios donde la información está distribuida espacialmente y existen múltiples objetivos en conflicto, no existe un único punto de vista privilegiado. Como en la naturaleza, la mejor solución se obtiene mediante la colaboración, y en el caso artificial, la metáfora más directa para implementar esta solución colectiva es mediante <b>sistemas multiagente</b>. Aquí, la inteligencia emerge del intercambio de información y la competencia regulada entre entidades autónomas, permitiendo abordar problemas de escala masiva que serían inmanejables para un controlador centralizado.
    </p>
</div>

## 1. Sistemas Multiagente

A pesar de la popularidad del término **agente**, no existe un acuerdo general acerca de su definición precisa. Algunos diccionarios, en un contexto genérico, lo definen como "*uno que actúa o tiene el poder o la autoridad para actuar... o representar a otro*". Una definición alternativa, proveniente de la psicología (Merani), es: *"lo que actúa o ejerce acción, por oposición al paciente, que la sufre"*. Esta segunda definición es particularmente interesante porque enfatiza el carácter proactivo y autónomo del agente: no reacciona pasivamente al mundo —lo transforma.

En el contexto computacional, las definiciones oscilan entre las dos definiciones anteriores, con diferencias dependiendo del contexto para el que se diseña, pero podríamos partir de una definición general en los siguientes términos:

> Un **agente** es una entidad computacional guiada por objetivos propios, capaz de percibir su entorno a través de **sensores** y actúar sobre él a través de **actuadores**.

<br>
<p align="center" style="margin-left:30px; margin-bottom: -15px;">
  <img src="images/The Matrix_Agents.jpg" width="800">
</p>

> "Son programas sintientes. Pueden moverse dentro y fuera de cualquier software que siga conectado a su sistema."
> Morpheus, The Matrix.

<br>

Esta definición puede hacerse más específica en términos de las propiedades que debiera poseer un sistema para ser considerado como un agente. Sin embargo, tampoco existe un acuerdo entre los diferentes autores, acerca de tal conjunto de características. En la literatura se suelen distinguir dos definciones en base a as características incluídas: 

La **definición blanda de agencia** que incluye un mínimo de características con las que la mayoría de los investigadores coincide:

- **Habilidad social:** capacidad de interactuar con otros agentes mediante comunicación directa o modificación del entorno compartido.
- **Reactividad:** percibe cambios en su entorno y responde a ellos en tiempo útil.
- **Orientación a objetivos:** su comportamiento está dirigido hacia metas, no es simplemente una respuesta refleja a estímulos.
- **Autonomía:** opera sin control externo directo; toma decisiones basadas en su propio estado interno.
- **Proactividad:** no solo reacciona —toma la iniciativa cuando su estado interno o sus objetivos lo requieren.
- **Continuidad temporal:** tiene existencia persistente; no es invocado y destruido,sino que mantiene un ciclo de vida propio.

Y la **definición dura** que incluye características más complejas como las sigientes:

- **Movilidad:** puede desplazarse entre nodos, sistemas o entornos de ejecución distintos.
- **Benevolencia:** actúa en beneficio de otros agentes o del sistema cuando no hay conflicto con sus propios objetivos.
- **Racionalidad:** selecciona las acciones que mejor satisfacen sus objetivos dado su modelo del mundo, evitando acciones que los contravengan.
- **Adaptabilidad:** modifica su comportamiento o su modelo interno en respuesta a cambios en el entorno o en los resultados de sus acciones.
- **Colaboración:** capaz de formar coaliciones, negociar y coordinar planes con otros agentes para alcanzar objetivos que ninguno lograría solo.

Alternativamente, podemos considerar estas dos definiciones no como categorías discretas sino como los extremos de una **línea de agencia**:

<p align="center">
  <img src="images/linea_de_agencia.png" width="600">
</p>

<br>

En cualquier punto de ese continuo, tres elementos estructurales definen a todo agente:

- **Percepción:** qué información del entorno puede observar (su visión local).
- **Deliberación:** cómo transforma esa percepción en una decisión —desde una regla simple hasta un modelo probabilístico complejo.
- **Acción:** cómo modifica el entorno o su propio estado como resultado de esa decisión.

### Agentes como Concepto de Software

Desde la perspectiva de la ingeniería de software, el agente puede entenderse como una evolución natural del encapsulamiento: cada paso en esta jerarquía delega más control hacia la entidad misma y menos hacia el código que la invoca.

---

#### Objeto — Encapsulamiento y abstracción

En la programación orientada a objetos, un *objeto* es una instancia de una clase que encapsula datos (atributos) y comportamientos (métodos). La idea central es organizar el código de forma modular y reutilizable, haciendo que cada objeto sea responsable de su propio estado.

**Características clave:**
- *Encapsulamiento:* agrupa datos y métodos relacionados, ocultando detalles internos y exponiendo solo lo necesario.
- *Abstracción:* representa entidades del mundo real mediante atributos y métodos.
- *Interacción:* los objetos se comunican entre sí a través de llamadas a métodos, sin modificar el estado interno del otro directamente.

El objeto es fundamentalmente **pasivo**: no hace nada a menos que otro componente lo invoque. Es una entidad inanimada —un contenedor de lógica que espera ser usado.



#### Actor — Concurrencia y persistencia

El *actor* introduce un ciclo de vida propio más allá de la simple invocación de métodos. Los actores operan de forma concurrente y se comunican exclusivamente mediante paso de mensajes asíncronos, sin memoria compartida. 

**Características clave:**
- *Concurrencia:* cada actor tiene su propio hilo de ejecución; no requiere sincronización explícita con otros actores.
- *Autonomía básica:* decide cuándo y cómo procesar los mensajes de su buzón, pero su comportamiento sigue siendo esencialmente reactivo —sin mensajes, el actor no actúa.
- *Persistencia temporal:* a diferencia del objeto, el actor existe de forma continua y puede mantener estado entre mensajes sucesivos.

El actor es ya una entidad **animada**: tiene vida propia, pero vive para responder. Su agencia está acotada al buzón de mensajes —no actúa por iniciativa propia.


#### Agente — Autonomía, deliberación y adaptabilidad

El *agente* representa la culminación de esta evolución. No solo responde a mensajes o eventos: tiene **objetivos internos** y actúa por iniciativa propia para alcanzarlos. Puede razonar sobre múltiples opciones, planificar secuencias de acciones y modificar su comportamiento en función de la experiencia.

**Características clave:**
- *Autonomía:* decide qué hacer en función de sus propios objetivos, sin depender de un controlador externo.
- *Deliberación:* evalúa alternativas antes de actuar, considerando consecuencias y estrategias —no solo reacciona al estímulo inmediato.
- *Adaptabilidad:* aprende de su entorno y de sus acciones pasadas, mejorando su comportamiento con el tiempo.
- *Interacción social:* colabora o compite con otros agentes para alcanzar objetivos que ninguno lograría en aislamiento.

El agente es una entidad con la que se puede **negociar**, no solo invocar. Este es el umbral conceptual relevante para los sistemas que estudiaremos: cuando el código deja de ser una herramienta que ejecutamos y se convierte en un participante que actúa y que interactúa con el usuario como componentes de un mismo ecoistema.


#### Comparativa

| Concepto | Control | Comportamiento | Iniciativa | Ejemplo |
|:---|:---|:---|:---|:---|
| **Objeto** | Externo total | Pasivo | Ninguna | `DataFrame`, `List` |
| **Actor** | Parcialmente propio | Reactivo (buzón) | Solo ante mensajes | Actor en Akka/Erlang |
| **Agente** | Autónomo | Pro-activo | Por objetivos propios | Partícula PSO, LLM Agent |


<br>
<p align="center"">
  <img src="images/objeto-actor-agente.png" width="600">
</p>


### Inteligencia Artificial Colectiva: SMA

Un **Sistema Multiagente (SMA)** es un sistema computacional compuesto por múltiples entidades autónomas —los **agentes**— que perciben un entorno compartido, toman decisiones dentro de su dominio de competencia y se articulan con otros agentes para resolver colectivamente un problema más amplio. Ningún agente necesita conocimiento global para cumplir su rol.

Esta arquitectura descansa en tres propiedades:

1. **Autonomía local.** Cada agente opera con información parcial del entorno y toma decisiones dentro de su dominio sin requerir autorización externa. Su comportamiento está gobernado por sus propios objetivos y estado interno.

2. **Interacción como mecanismo de cómputo.** Los agentes se comunican, cooperan, compiten o delegan. La solución al problema no reside en ningún agente individual: es el producto de sus intercambios.

3. **Emergencia.** El comportamiento colectivo exhibe propiedades que no están presente en ninguno de sus agentes por separado. El sistema resuelve lo que sus partes, en aislamiento, ni siquiera pueden formular.

El comportamiento emergente es particularmente valioso en ciencia de datos porque los problemas que enfrentamos rara vez tienen una solución analítica cerrada —y cuando la tienen, rara vez es única. Aquí es donde los SMA ofrecen una ventaja estructural.

El **No Free Lunch Theorem** establece que ningún algoritmo domina a todos los demás en todos los problemas: lo que gana en un paisaje de búsqueda, lo pierde en otro. La respuesta clásica a este teorema es elegir el mejor algoritmo disponible para el problema en cuestión. La respuesta de los SMA es más ambiciosa: **no elegir uno, sino desplegar varios simultáneamente**.

Un SMA permite combinar en un mismo sistema agentes con estrategias radicalmente distintas —un predictor bayesiano, un árbol de decisión, un optimizador evolutivo, un modelo de series de tiempo— cada uno resolviendo el problema desde su propia perspectiva, con sus propias fortalezas. Los agentes árbitro integran esas perspectivas y producen una solución que ninguna de las partes podría alcanzar sola.

Esta es la intuición detrás de los métodos ensemble, del aprendizaje federado y de los sistemas de optimización híbrida: no hay un algoritmo perfecto, pero sí hay sociedades de algoritmos que se complementan. El SMA es el marco que hace esa sociedad diseñable, no accidental.

### Ejercicio: Predicción de series de tiempo mediante un SMA

Para entender qué hay "dentro" de un Sistema Multiagente, implementaremos una simulación donde tres **agentes especialistas** predicen el siguiente valor de una serie de tiempo usando estrategias radicalmente distintas —estadística clásica, memoria por similitud y aprendizaje no lineal— y un **agente coordinador** integra sus propuestas en tiempo real.

<p align="center"">
  <img src="images/sma_predictores_pizarra.png" width="600">
</p>

El ejercicio está diseñado para hacer visibles tres propiedades del SMA que
hasta ahora hemos descrito solo conceptualmente:

- **Autonomía local:** cada agente corre en su propio hilo, decide cuándo
  está listo y deposita su propuesta sin coordinación explícita con los demás.
- **Asincronía como propiedad del sistema:** los agentes no llegan al mismo
  tiempo. El coordinador no espera al más lento: actúa con quórum parcial
  y avanza. Los rezagados se sincronizan al siguiente paso.
- **Emergencia:** el dictamen final no es el mejor agente individual —es la
  mediana de las propuestas disponibles, una decisión que ningún agente
  podría tomar solo porque ninguno conoce las predicciones de los demás.

#### Arquitectura

El sistema implementa el patrón **Blackboard**: un entorno compartido donde
los agentes depositan propuestas y el coordinador las lee para producir
dictámenes. Es uno de los patrones de coordinación más usados en SMA
jerárquicos.

In [14]:
import threading
import time
import warnings
import numpy as np

# ──────────────────────────────────────────────────────────────────────────────────────────
# SERIE — Oscilador amortiguado con ruido gaussiano
# Estructura temporal clara —amplitud decreciente + oscilación periódica— lo que hace que 
# ARIMA, KNN y MLP lleguen a predicciones distintas y comparables. Es predecible a corto 
# plazo pero no trivial, a diferencia de una serie lineal.
# ──────────────────────────────────────────────────────────────────────────────────────────
def generar_serie(n_obs=120, seed=42):
    np.random.seed(seed)
    t     = np.linspace(0, 12 * np.pi, n_obs)
    serie = np.exp(-0.05 * t) * np.sin(t) + np.random.normal(0, 0.05, n_obs)
    return serie

# ──────────────────────────────────────────────────────────────────────────────────────────
# PIZARRA
# ──────────────────────────────────────────────────────────────────────────────────────────
class Blackboard:
    def __init__(self, serie_inicial, n_pasos=5, quorum=2):
        self.serie_base  = list(serie_inicial)
        self.n_pasos     = n_pasos
        self.quorum      = quorum
        self.lock        = threading.Lock()
        self.t0          = time.time()

        # Diccionario que registra las propuestas de cada agente para ese paso. 
        # El coordinador lo monitorea.
        self.pizarra     = {p: {} for p in range(n_pasos)}               
        self.dictamen    = {}

        # Serie original extendida con los dictámenes del coordinador. 
        # Es la entrada que reciben los agentes en cada paso siguiente.
        self.serie_ext   = list(serie_inicial)

        # `threading.Event` por paso. El coordinador lo activa cuando publica el 
        # dictamen; los agentes están bloqueados en `.wait()` hasta ese momento. 
        # Sin este mecanismo, un agente rápido podría avanzar al paso 3 antes de 
        # que el coordinador haya procesado el paso 2.
        self.paso_listo  = {p: threading.Event() for p in range(n_pasos)}
        self.terminado   = threading.Event()

    # Método para escribir propuestas individuales en la pizarra. 
    # Cada agente se identifica y publica su propuesta.
    def depositar(self, nombre, paso, valor):
        ts = time.time() - self.t0
        with self.lock: # El segundo hilo espera en la puerta hasta que el primero 
                        # termina de escribir y sale del bloque. Evita carrera por
                        # recursos. 
            self.pizarra[paso][nombre] = valor
            n = len(self.pizarra[paso])
            print(f"  t+{ts:5.2f}s │ paso {paso+1}/{self.n_pasos} │ "
                  f"{nombre:<18} │ pred: {valor:+.4f}  [{n} en pizarra]")

    # Método para leer de la pizarra.
    # Regresa el conjunto de propuestas para ese paso.
    def obtener_paso(self, paso):
        with self.lock: # defensive programming :-)
            return dict(self.pizarra[paso])

    # Método para escribir la solución integrada de cada paso.
    # Lo utiliza el agente coordinador.
    def registrar_dictamen(self, paso, valor, agentes_usados):
        ts = time.time() - self.t0
        with self.lock: # defensive programming :-)
            self.dictamen[paso] = (valor, agentes_usados)
            self.serie_ext.append(valor)
        self.paso_listo[paso].set()
        print(f"  {'─'*75}")
        print(f"  t+{ts:5.2f}s │ DICTAMEN paso {paso+1}/{self.n_pasos} │ "
              f"mediana={valor:+.4f} │ usando: {agentes_usados}")
        print(f"  {'─'*75}\n") 

    # Semáforo de paso. Determina cuando se ha registrado un nuevo valor.
    # El agente predictor necesita esta valor para continuar.
    def esperar_dictamen(self, paso, timeout=30):
        self.paso_listo[paso].wait(timeout=timeout)
        with self.lock: # defensive programming :-)
            return self.serie_ext[:]

# ──────────────────────────────────────────────────────────────────────────────────────────
# AGENTES CONSULTORES
# Clase que hereda de threading.Thread, es decir, los objetos se ejecutan sobre su propio
# hilo. En un sistema multicore, cada hilo puede utilizar un core distinto del procesador.
# ... si estuviera escrito en Java :-) Hasta la fecha, Python sólo simula threading.
# ──────────────────────────────────────────────────────────────────────────────────────────
class AgenteConsultor(threading.Thread):
    def __init__(self, nombre, entorno, latencia_base):
        # daemon=True especifica que este hilo se ejecuta en segundo plano y se cierra
        # automáticamente cuando el programa principal finaliza.
        super().__init__(name=nombre, daemon=True)
        self.nombre        = nombre
        self.entorno       = entorno
        self.latencia_base = latencia_base

    # Esta es una clase abstracta implícita...
    def predecir_paso(self, serie):
        raise NotImplementedError # aquí falla en tiempo de ejecución

    # .run() define la actividad que se ejecutará en un hilo. 
    # No se debe llamar directamente, para ello se utiliza .start() 
    def run(self): 
        print(f"  {self.nombre} iniciando...\n")
        for paso in range(self.entorno.n_pasos):
            serie = (self.entorno.esperar_dictamen(paso - 1)
                     if paso > 0
                     else self.entorno.serie_base[:])

            # El agente se envía a dormir con una componente aleatorio para simular
            # la variabilidad de carga del sistema. El orden de llegada es impredecible
            # entre ejecuciones, como ocurre en un sistema distribuido real.
            time.sleep(self.latencia_base + np.random.uniform(0.1, 0.5))
            valor = self.predecir_paso(serie)
            self.entorno.depositar(self.nombre, paso, valor)

# Clase derivada para definir un tipo de AgenteConsultor.
class AgenteARIMA(AgenteConsultor):
    # Sólo implementa su propio método `predecir_paso`
    def predecir_paso(self, serie):
        from statsmodels.tsa.arima.model import ARIMA
        with warnings.catch_warnings():
            warnings.simplefilter("ignore")
            modelo = ARIMA(serie, order=(2, 1, 1)).fit()
        # forecast() devuelve ndarray en versiones recientes — usamos [0]
        resultado = modelo.forecast(steps=1)
        return float(resultado[0])


class AgenteKNN(AgenteConsultor):
    def predecir_paso(self, serie, k=5, window=10):
        from sklearn.neighbors import KNeighborsRegressor
        X, y = [], []
        for i in range(len(serie) - window):
            X.append(serie[i:i+window])
            y.append(serie[i+window])
        if len(X) < k:
            return float(serie[-1])
        modelo = KNeighborsRegressor(n_neighbors=k).fit(np.array(X), np.array(y))
        return float(modelo.predict([serie[-window:]])[0])


class AgenteMLP(AgenteConsultor):
    def predecir_paso(self, serie, window=10):
        from sklearn.neural_network import MLPRegressor
        with warnings.catch_warnings():
            warnings.simplefilter("ignore")   # convergencia parcial aceptable en 1 paso
            X, y = [], []
            for i in range(len(serie) - window):
                X.append(serie[i:i+window])
                y.append(serie[i+window])
            if len(X) < 10:
                return float(serie[-1])
            modelo = MLPRegressor(
                hidden_layer_sizes=(16, 8),
                max_iter=2000,
                random_state=42,
                tol=1e-3
            ).fit(np.array(X), np.array(y))
        return float(modelo.predict([serie[-window:]])[0])


# ─────────────────────────────────────────────────────────────
# AGENTE COORDINADOR
# Este agente integra las propuestas individuales utilizando la mediana.
# No espera a todos los agentes: en cuanto `quorum=2` agentes depositan su propuesta, 
# el coordinador calcula la mediana y publica el dictamen.
# ─────────────────────────────────────────────────────────────
class AgenteCoordinador(threading.Thread):
    def __init__(self, entorno):
        super().__init__(name="Coordinador", daemon=True)
        self.entorno = entorno

    def run(self):
        for paso in range(self.entorno.n_pasos):
            while True:
                snap = self.entorno.obtener_paso(paso)
                if len(snap) >= self.entorno.quorum:
                    valores   = list(snap.values())
                    dictamen  = float(np.median(valores))
                    self.entorno.registrar_dictamen(paso, dictamen, list(snap.keys()))
                    break
                time.sleep(0.2)
        self.entorno.terminado.set()


# ─────────────────────────────────────────────────────────────
# TABLA RESUMEN
# ─────────────────────────────────────────────────────────────
def imprimir_tabla_final(entorno):
    print("\n" + "═"*80)
    print("  PIZARRA FINAL — resumen por paso")
    print("═"*80)
    print(f"  {'Paso':<6} {'ARIMA':>12} {'KNN':>12} {'MLP':>12} "
          f"{'Dictamen':>12}  Agentes usados")
    print(f"  {'─'*75}")
    for paso in range(entorno.n_pasos):
        fila          = entorno.pizarra[paso]
        dic_val, ag   = entorno.dictamen.get(paso, (float('nan'), []))
        arima  = f"{fila.get('ARIMA', float('nan')):+.4f}"
        knn    = f"{fila.get('KNN',   float('nan')):+.4f}"
        mlp    = f"{fila.get('MLP',   float('nan')):+.4f}"
        dic_s  = f"{dic_val:+.4f}"
        usados = ", ".join(ag)
        print(f"  {paso+1:<6} {arima:>12} {knn:>12} {mlp:>12} "
              f"{dic_s:>12}  ← {usados}")
    print("═"*80)


# ─────────────────────────────────────────────────────────────
# EJECUCIÓN
# ─────────────────────────────────────────────────────────────
def iniciar_consultoria(n_pasos=5, quorum=2):
    print("═"*80)
    print("  SISTEMA MULTIAGENTE — Predicción paso a paso")
    print("  Serie: oscilador amortiguado  │  "
          f"pasos={n_pasos}  quórum={quorum}/3")
    print("═"*80 + "\n")

    serie       = generar_serie(n_obs=120)
    bb          = Blackboard(serie, n_pasos=n_pasos, quorum=quorum)

    # Creación de los agentes... como cualquier objeto
    agentes     = [
        AgenteARIMA("ARIMA", bb, latencia_base=1.2),
        AgenteKNN  ("KNN",   bb, latencia_base=0.5),
        AgenteMLP  ("MLP",   bb, latencia_base=0.8),
    ]
    coordinador = AgenteCoordinador(bb)

    # Lanzamiento de los agentes
    coordinador.start() # Crea un nuevo hilo, le asigna recursos y ejecuta run()
    for a in agentes:
        a.start()

    bb.terminado.wait(timeout=120)
    imprimir_tabla_final(bb)

iniciar_consultoria(n_pasos=5, quorum=2)

════════════════════════════════════════════════════════════════════════════════
  SISTEMA MULTIAGENTE — Predicción paso a paso
  Serie: oscilador amortiguado  │  pasos=5  quórum=2/3
════════════════════════════════════════════════════════════════════════════════

  ARIMA iniciando...

  KNN iniciando...

  MLP iniciando...

  t+ 0.90s │ paso 1/5 │ KNN                │ pred: +0.0783  [1 en pizarra]
  t+ 1.08s │ paso 1/5 │ MLP                │ pred: +0.1207  [2 en pizarra]
  ───────────────────────────────────────────────────────────────────────────
  t+ 1.22s │ DICTAMEN paso 1/5 │ mediana=+0.0995 │ usando: ['KNN', 'MLP']
  ───────────────────────────────────────────────────────────────────────────

  t+ 1.50s │ paso 1/5 │ ARIMA              │ pred: +0.1121  [3 en pizarra]
  t+ 2.08s │ paso 2/5 │ KNN                │ pred: +0.1099  [1 en pizarra]
  t+ 2.39s │ paso 2/5 │ MLP                │ pred: +0.0222  [2 en pizarra]
  ─────────────────────────────────────────────────────────────────

---

<br>

**Análisis de los resultados**

Observemos los resultados del primer paso

```
  t+ 0.91s │ paso 1/5          │ KNN             │ pred: +0.0783  [1 en pizarra]
  t+ 1.09s │ paso 1/5          │ MLP             │ pred: +0.1207  [2 en pizarra]
─────────────────────────────────────────────────────────────────────────────────
  t+ 1.22s │ DICTAMEN paso 1/5 │ mediana=+0.0995 │ usando: ['KNN', 'MLP']
─────────────────────────────────────────────────────────────────────────────────

  t+ 1.49s │ paso 1/5          │ ARIMA           │ pred: +0.1121  [3 en pizarra]
  t+ 2.08s │ paso 2/5          │ KNN             │ pred: +0.1099  [1 en pizarra]
  ...
```

ARIMA llega 0.58 segundos después que KNN —el coordinador ya publicó el dictamen del paso 1 antes de que ARIMA terminara. La predicción de ARIMA queda registrada en la pizarra pero **no influyó en ese dictamen**. Esto es quórum parcial: el sistema avanza sin esperar al más lento. Este retraso es la consecuencia de que ARIMA re-ajusta un modelo estadístico completo en cada paso (estimación de parámetros por máxima verosimilitud), mientras KNN y MLP simplemente buscan vecinos o propagan hacia adelante. El costo computacional es real y observable en los timestamps.

**La tabla final**

```
Paso    ARIMA      KNN      MLP   Dictamen  Agentes usados
1      +0.1121  +0.0783  +0.1207  +0.0995  ← KNN, MLP
2      +0.1319  +0.1099  +0.0222  +0.0660  ← KNN, MLP
3      +0.0771  +0.1263  +0.1251  +0.1257  ← KNN, MLP
4         nan   +0.1536  +0.1303  +0.1420  ← KNN, MLP
5         nan   +0.1807  +0.1660  +0.1734  ← KNN, MLP
```

Tres lecturas de esta tabla:

**1. Divergencia entre agentes.** En el paso 2, MLP predice `+0.0222` mientras ARIMA predice `+0.1319` —una diferencia de 0.11 sobre valores en el rango [0, 0.2]. Cada agente ve la misma serie pero llega a conclusiones distintas porque su estrategia se basa en un modelo del mundo diferente. Esto es exactamente la **diversidad de perspectivas** que hace valiosa la arquitectura multiagente.

**2. El coordinador como árbitro robusto.** La mediana de paso 2 es `+0.0660`, más cercana a KNN que a cualquiera de los otros dos. No es un promedio: es la propuesta central que minimiza la influencia de los extremos. En paso 3, donde KNN y MLP coinciden casi exactamente (`+0.1263` y `+0.1251`), el dictamen refleja ese consenso (`+0.1257`).

**3. `nan` en ARIMA pasos 4 y 5.** ARIMA terminó de procesar el paso 3 en `t+4.58s`, cuando el coordinador ya había publicado el dictamen del paso 4 en `t+4.88s`. Para el paso 5 ocurrió lo mismo. La pizarra registra `nan` porque ARIMA nunca llegó a tiempo —pero el sistema funcionó correctamente con los dos agentes restantes. Esto ilustra una propiedad deseable de los SMA: **tolerancia a fallos parciales**. El sistema no colapsa cuando un agente se rezaga o falla; continúa con la información disponible.

<div style="margin-top: 30px; padding: 10px; background: linear-gradient(135deg, #24398A 0%, #1a2a66 100%); box-shadow: 0 8px 16px rgba(36, 57, 138, 0.3); color: #ffffaa">
<h3> Tarea: Evaluación difusa de riesgo crediticio</h3>
</div>

#### Objetivo

Construir y analizar un sistema de inferencia colectiva para estimar el **riesgo crediticio**, utilizando datos reales del dataset  
**German Credit Dataset** (UCI Machine Learning Repository).

El dataset contiene información de solicitantes de crédito, incluyendo:

- duración del crédito  
- monto del préstamo  
- historial crediticio  
- empleo  
- nivel de ahorro  
- estado personal  

Variable objetivo: **Riesgo**

#### Tareas:

- Crear agentes con modelos diferentes modelos difusos (Mamdani, Tsukamoto, TSK) para resolver el problema
- Generar la solución global. **Bonus**: Generar una solución que tome en cuenta la reputación como analista de cada agente.
- Discutir resultados. 

#### Entregables

- Notebook reproducible  
- Tabla comparativa (diferentes modelos contra solución teórica y contra solución global)
- Discusión técnica  


<div style="border: 2px solid #24398A; border-radius: 15px; padding: 20px; background-color: #fcfcfc;">
    <h3 style="color: #24398A; margin-top: 0;">Conclusiones: Inteligencia Distribuida</h3>
    <p style="font-size: 15px; line-height: 1.6;">
        Los Sistemas Multiagente (SMA) representan una arquitectura descentralizada para la resolución de problemas. La transición de una inteligencia centralizada a una colectiva permite modelar sistemas dinámicos donde la interacción local define el comportamiento global.
    </p>
    <ul style="font-size: 14px; color: #333;">
        <li><b>Autonomía y Adaptabilidad:</b> Cada agente actúa bajo reglas locales, lo que otorga al sistema una resiliencia superior ante fallos puntuales.</li>
        <li><b>Aplicación en Big Data:</b> Los SMA facilitan la descomposición de tareas complejas en subproblemas manejables y distribuidos.</li>
    </ul>
    
</div>